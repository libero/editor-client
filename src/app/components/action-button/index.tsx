import React from 'react';
import { Button, PropTypes } from '@material-ui/core';
import { mapValues } from 'lodash';
import AddIcon from '@material-ui/icons/Add';

import { useActionButtonStyles } from './styles';

type MuiButtonVariant = 'text' | 'outlined' | 'contained';

const VARIANTS = {
  containedWarning: {
    variant: 'contained' as MuiButtonVariant,
    classes: {
      root: 'containedRedButton'
    }
  },
  outlinedWarning: {
    variant: 'outlined' as MuiButtonVariant,
    classes: {
      root: 'outlinedRedButton'
    }
  },
  primaryContained: {
    variant: 'contained' as MuiButtonVariant,
    color: 'primary',
    classes: {}
  },
  secondaryOutlined: {
    variant: 'outlined',
    classes: {}
  },
  addEntity: {
    size: 'small',
    classes: { root: 'addEntityButton_root', startIcon: 'addEntityButton_icon' },
    variant: 'text',
    startIcon: <AddIcon />
  }
} as {
  [key: string]: {
    variant: MuiButtonVariant;
    classes: Record<string, string>;
    color?: PropTypes.Color;
  };
};

export type VariantType = keyof typeof VARIANTS;

interface ActionButtonProps {
  onClick: () => void;
  title: string;
  variant: VariantType;
  className?: string;
}

export const ActionButton: React.FC<ActionButtonProps> = (props) => {
  const classes = useActionButtonStyles();
  const variantProps = VARIANTS[props.variant];
  const variantClasses = mapValues(variantProps.classes, (className: string) => classes[className]);
  const buttonProps = { ...variantProps, classes: variantClasses };
  return (
    <Button onClick={props.onClick} {...buttonProps} className={props.className}>
      {props.title}
    </Button>
  );
};
