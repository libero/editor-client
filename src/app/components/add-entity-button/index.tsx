import React from 'react';
import AddIcon from '@material-ui/icons/Add';
import { Button } from '@material-ui/core';
import { useAddEntityBtnStyles } from './styles';

export const AddEntityButton: React.FC = () => {
  const classes = useAddEntityBtnStyles();
  return (
    <Button
      size="small"
      classes={{ root: classes.root, startIcon: classes.icon }}
      variant="text"
      startIcon={<AddIcon />}
    >
      Add Author
    </Button>
  );
};
