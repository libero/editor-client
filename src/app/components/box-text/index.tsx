import React from 'react';
import { NodeView, EditorView } from 'prosemirror-view';
import { Node as ProsemirrorNode } from 'prosemirror-model';
import ReactDOM from 'react-dom';
import { ThemeProvider } from '@material-ui/core/styles';
import { NodeSelection } from 'prosemirror-state';

import { BoxTextEditor, BoxTextEditorHandle } from './box-text-editor';
import { theme } from '../../styles/theme';
import { NodeViewContext } from '../../utils/view.utils';

export class BoxTextNodeView implements NodeView {
  dom?: HTMLElement;
  boxTextEditor: React.RefObject<BoxTextEditorHandle>;

  constructor(private node: ProsemirrorNode, private view: EditorView, private getPos: () => number) {
    this.dom = document.createElement('section');
    this.boxTextEditor = React.createRef();
    ReactDOM.render(
      <ThemeProvider theme={theme}>
        <NodeViewContext.Provider
          value={{
            view: this.view,
            getPos: this.getPos,
            getNode: () => this.node
          }}
        >
          <BoxTextEditor ref={this.boxTextEditor} node={this.node} onDelete={this.handleDelete} />
        </NodeViewContext.Provider>
      </ThemeProvider>,
      this.dom
    );

    this.dom.addEventListener('drop', this.handleDrop, true);
  }

  handleDrop = (event: Event): void => {
    event.stopPropagation();
    event.preventDefault();
  };

  handleDelete = (): void => {
    const change = this.view.state.tr
      .setSelection(new NodeSelection(this.view.state.doc.resolve(this.getPos())))
      .deleteSelection();
    this.view.dispatch(change);
  };

  update(node: ProsemirrorNode): boolean {
    this.node = node;
    this.boxTextEditor.current.updateContent(node);
    return true;
  }

  stopEvent(evt): boolean {
    return this.dom.contains(evt.target) || /drag/.test(evt.type);
  }

  destroy(): void {
    this.dom.removeEventListener('drop', this.handleDrop, true);
  }

  ignoreMutation(): boolean {
    return true;
  }
}
