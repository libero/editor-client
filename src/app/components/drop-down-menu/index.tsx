import React, { useCallback, useRef } from 'react';
import { Popper, ClickAwayListener, Paper, MenuList, Button, MenuItem } from '@material-ui/core';
import ArrowDropDownIcon from '@material-ui/icons/ArrowDropDown';

import { useDropDownStyles } from './styles';

export interface DropDownMenuItemProps {
  title: string;
  enabled: boolean;
  selected?: boolean;
  action(): void;
}

export interface DropDownMenuProps {
  title: string;
  entries: DropDownMenuItemProps[];
}

export const DropDownMenu: React.FC<DropDownMenuProps> = ({ title, entries }) => {
  const [open, setOpen] = React.useState(false);
  const menuRef = useRef();
  const handleOpenMenuClick = useCallback(() => setOpen(true), [setOpen]);
  const handleMenuClose = useCallback(() => setOpen(false), [setOpen]);
  const classes = useDropDownStyles();

  const handleMenuItemClick = useCallback((action) => {
    if (action) {
      action();
    }
    setOpen(false);
  }, []);

  const menuItems = entries.map((entry, index) => {
    return (
      <MenuItem
        disabled={!entry.enabled}
        key={index}
        selected={entry.selected}
        onClick={handleMenuItemClick.bind(null, entry.action)}
      >
        {entry.title}
      </MenuItem>
    );
  });

  return (
    <div>
      <Button ref={menuRef} aria-haspopup="true" onClick={handleOpenMenuClick} className={classes.button}>
        {title}
        <ArrowDropDownIcon />
      </Button>
      <Popper anchorEl={menuRef.current} role={undefined} transition disablePortal open={open}>
        <Paper>
          <ClickAwayListener onClickAway={handleMenuClose}>
            <MenuList autoFocusItem={open}>{menuItems}</MenuList>
          </ClickAwayListener>
        </Paper>
      </Popper>
    </div>
  );
};
