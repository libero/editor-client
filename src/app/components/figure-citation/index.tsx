import React from 'react';
import ReactDOM from 'react-dom';
import { Node as ProsemirrorNode } from 'prosemirror-model';
import { EditorView, NodeView } from 'prosemirror-view';
import { ThemeProvider } from '@material-ui/core/styles';

import { theme } from '../../styles/theme';
import { FigureCitationEditorPopup, FiguresListEntry, UNLABELLED_FIGURE_TEXT } from './figure-citation-editor-popup';
import { TextSelection } from 'prosemirror-state';
import { store } from '../../store';
import { Provider } from 'react-redux';

export class FigureCitationNodeView implements NodeView {
  dom?: HTMLAnchorElement;
  contentDOM?: HTMLSpanElement;
  figureEditorContainer: HTMLDivElement;

  constructor(private node: ProsemirrorNode, private view: EditorView, private getPos: () => number) {
    this.dom = document.createElement('a');
    this.dom.style.cursor = 'pointer';
    this.dom.addEventListener('click', this.selectNode);
    this.contentDOM = document.createElement('span');
    this.dom.appendChild(this.contentDOM);
    if (this.node.attrs.figIds === null) {
      this.open();
    }
  }

  selectNode = (): void => {
    this.dom.classList.add('ProseMirror-selectednode');
    this.open();
  };

  deselectNode(): void {
    this.dom.classList.remove('ProseMirror-selectednode');
    this.close();
  }

  update(node: ProsemirrorNode): boolean {
    this.node = node;
    return true;
  }

  open(): void {
    this.figureEditorContainer = this.view.dom.parentNode.appendChild(document.createElement('div'));

    ReactDOM.render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <FigureCitationEditorPopup
            selectedIds={this.node.attrs.figIds || []}
            anchorEl={this.dom}
            onClose={this.close}
            onChange={this.handleChange}
          />
        </ThemeProvider>
      </Provider>,
      this.figureEditorContainer
    );
  }

  stopEvent(): boolean {
    return true;
  }

  ignoreMutation(): boolean {
    return true;
  }

  close = (): void => {
    this.dom.classList.remove('ProseMirror-selectednode');
    if (this.figureEditorContainer) {
      ReactDOM.unmountComponentAtNode(this.figureEditorContainer);
      this.figureEditorContainer.parentNode.removeChild(this.figureEditorContainer);
      this.figureEditorContainer = null;
    }
  };

  handleChange = (figures: FiguresListEntry[]): void => {
    const ids = figures.map(({ id }) => id);
    const change = this.view.state.tr.setNodeMarkup(this.getPos(), null, { figIds: ids });

    if (this.node.textContent === '???' && figures.length > 0) {
      const nodeContent = (figures[0].name || UNLABELLED_FIGURE_TEXT).replace(/\.$/, '');
      change
        .setSelection(TextSelection.create(change.doc, this.getPos() + 1, this.getPos() + this.node.nodeSize - 1))
        .replaceSelectionWith(this.view.state.schema.text(nodeContent));
    } else if (figures.length === 0) {
      change
        .setSelection(TextSelection.create(change.doc, this.getPos() + 1, this.getPos() + this.node.nodeSize - 1))
        .replaceSelectionWith(this.view.state.schema.text('???'));
    }

    this.view.dispatch(change);
  };
}
