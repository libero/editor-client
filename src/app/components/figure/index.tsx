import React from 'react';
import { NodeView, EditorView } from 'prosemirror-view';
import { Node as ProsemirrorNode } from 'prosemirror-model';
import ReactDOM from 'react-dom';
import { ThemeProvider } from '@material-ui/core/styles';
import { Transaction, TextSelection, NodeSelection } from 'prosemirror-state';

import { StepMap } from 'prosemirror-transform';

import { theme } from 'app/styles/theme';
import { FigureEditor, FigureEditorHandle } from 'app/components/figure/figure-editor';

export class FigureNodeView implements NodeView {
  dom?: HTMLElement;
  figureEditor: React.RefObject<FigureEditorHandle>;

  constructor(
    private node: ProsemirrorNode,
    private view: EditorView,
    private getPos: () => number,
    private isContainerActive: () => boolean
  ) {
    this.dom = document.createElement('section');
    this.figureEditor = React.createRef();
    ReactDOM.render(
      <ThemeProvider theme={theme}>
        <FigureEditor
          ref={this.figureEditor}
          figureNode={this.node}
          onDelete={this.handleDelete}
          onLabelChange={this.handleLabelChange}
          onImageChange={this.handleImageChange}
          onNodeChange={this.handleContentChange}
          onSelectionChange={this.handleSelectionChange}
        />
      </ThemeProvider>,
      this.dom
    );
  }

  handleLabelChange = (label: string) => {
    const newAttributes = { label, img: this.node.attrs.img };
    const change = this.view.state.tr.setNodeMarkup(this.getPos(), null, newAttributes);
    this.view.dispatch(change);
  };

  handleImageChange = (img: string) => {
    const newAttributes = { img, label: this.node.attrs.label };
    const change = this.view.state.tr.setNodeMarkup(this.getPos(), null, newAttributes);
    this.view.dispatch(change);
  };

  handleContentChange = (nodeViewChange: Transaction, offset: number) => {
    this.view.dispatch(this.transformNodeViewChanges(nodeViewChange, offset));
  };

  handleDelete = () => {
    const change = this.view.state.tr
      .setSelection(new NodeSelection(this.view.state.doc.resolve(this.getPos())))
      .deleteSelection();
    this.view.dispatch(change);
  };

  handleSelectionChange = (anchor: number, head: number) => {
    const offset = this.getPos() + 1;
    const start = Math.min(anchor, head);
    const end = Math.max(anchor, head);
    const selection = TextSelection.create(this.view.state.doc, start + offset, end + offset);
    if (!selection.eq(this.view.state.selection)) {
      const selectionChange = this.view.state.tr.setSelection(selection);
      this.view.dispatch(selectionChange);
      this.view.dom.dispatchEvent(new FocusEvent('focus'));
    }
  };

  update(node: ProsemirrorNode) {
    if (!node.sameMarkup(this.node)) {
      this.node = node;
    }
    this.figureEditor.current.updateContent(node);
    if (this.isContainerActive() && !this.figureEditor.current.hasFocus()) {
      this.figureEditor.current.focusFromSelection(this.view.state.selection, this.getPos());
    }
    return true;
  }

  stopEvent(evt) {
    return this.dom.contains(evt.target);
  }

  ignoreMutation() {
    return true;
  }

  private transformNodeViewChanges(change: Transaction, additionalOffset: number): Transaction {
    const parentChange = this.view.state.tr;
    const offsetMap = StepMap.offset(this.getPos() + 2 + additionalOffset);
    const steps = change.steps;
    for (let i = 0; i < steps.length; i++) {
      parentChange.step(steps[i].map(offsetMap));
    }
    return parentChange;
  }
}
