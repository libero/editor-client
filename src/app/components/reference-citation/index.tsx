import React from 'react';
import ReactDOM from 'react-dom';
import { v4 as uuidv4 } from 'uuid';
import { Node as ProsemirrorNode } from 'prosemirror-model';
import { EditorView, NodeView } from 'prosemirror-view';
import { Provider } from 'react-redux';
import { ThemeProvider } from '@material-ui/core/styles';

import { TextSelection } from 'prosemirror-state';

import { theme } from 'app/styles/theme';
import { getRefNodeText, Reference } from 'app/models/reference';
import { store } from 'app/store';
import { ReferenceCitationEditorPopup } from 'app/components/reference-citation/reference-citation-editor-popup';

export class ReferenceCitationNodeView implements NodeView {
  dom?: HTMLAnchorElement;
  refEditorContainer: HTMLDivElement;

  constructor(private node: ProsemirrorNode, private view: EditorView, private getPos) {
    this.dom = document.createElement('a');
    this.dom.style.cursor = 'pointer';
    this.dom.textContent = this.node.attrs.refText || '???';
    this.dom.addEventListener('click', this.selectNode);
    if (this.node.attrs.refId === null) {
      this.open();
    }
  }

  selectNode = (): void => {
    this.dom.classList.add('ProseMirror-selectednode');
    this.open();
  };

  deselectNode(): void {
    this.dom.classList.remove('ProseMirror-selectednode');
    this.close();
  }

  open(): void {
    this.refEditorContainer = this.view.dom.parentNode.appendChild(document.createElement('div'));
    this.refEditorContainer.style.position = 'absolute';
    this.refEditorContainer.style.zIndex = '10';

    ReactDOM.render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <ReferenceCitationEditorPopup
            anchorEl={this.dom}
            editorView={this.view}
            node={this.node}
            onClose={this.close}
            onChange={this.handleChange}
          />
        </ThemeProvider>
      </Provider>,
      this.refEditorContainer
    );
  }

  stopEvent(): boolean {
    return true;
  }

  ignoreMutation(): boolean {
    return true;
  }

  close = (): void => {
    this.dom.classList.remove('ProseMirror-selectednode');
    if (this.refEditorContainer) {
      ReactDOM.unmountComponentAtNode(this.refEditorContainer);
      this.refEditorContainer.parentNode.removeChild(this.refEditorContainer);
      this.refEditorContainer = null;
    }
  };

  handleChange = (ref: Reference): void => {
    const attrs = ref
      ? { refId: ref.id || uuidv4(), refText: getRefNodeText(ref) }
      : { refId: undefined, refText: undefined };

    const schema = this.view.state.schema;
    const change = this.view.state.tr.replaceWith(
      this.getPos(),
      this.getPos() + this.node.nodeSize,
      schema.nodes['refCitation'].create(attrs)
    );
    // due to browser managing cursor position on focus and blur the cursor is sometimes reset to 0
    // to rectify this behaviour we move cursor back to before the citation
    change.setSelection(new TextSelection(change.doc.resolve(this.getPos())));
    this.view.dispatch(change);
    this.close();
  };
}
