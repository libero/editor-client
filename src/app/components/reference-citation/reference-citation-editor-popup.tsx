import React, { useCallback, SyntheticEvent, useState } from 'react';
import { EditorView } from 'prosemirror-view';
import { useSelector, useDispatch } from 'react-redux';
import { Node as ProsemirrorNode } from 'prosemirror-model';
import { TextField, InputAdornment, Popper, Paper, ClickAwayListener } from '@material-ui/core';
import CheckCircleIcon from '@material-ui/icons/CheckCircle';
import Interweave from 'interweave';
import AddIcon from '@material-ui/icons/Add';
import ClearIcon from '@material-ui/icons/Clear';
import { has, get } from 'lodash';

import { getReferences } from 'app/selectors/manuscript.selectors';
import { getRefListAuthorsNames, Reference } from 'app/models/reference';
import { useReferenceEditorStyles } from 'app/components/reference-citation/styles';
import * as manuscriptActions from 'app/actions/manuscript.actions';
import { stringifyEditorState } from 'app/utils/view.utils';
import { ReactFCProps } from 'app/utils/types';
import { ReferenceFormDialog } from 'app/containers/reference-form-dialog/reference-form-dialog';
import { ModalContainer } from 'app/containers/modal-container';

interface ReferenceCitationEditorPopupProps {
  editorView: EditorView | undefined;
  onClose(): void;
  anchorEl: HTMLElement;
  onChange(ref: Reference): void;
  node?: ProsemirrorNode;
}

const renderReferenceModal = (props: ReactFCProps<typeof ReferenceFormDialog>) => {
  return <ModalContainer title={'Reference'} params={props} component={ReferenceFormDialog} />;
};

const getRefListItemText = (ref: Reference) => {
  return [
    `${getRefListAuthorsNames(ref)}, ${get(ref.referenceInfo, 'year')}`,
    has(ref.referenceInfo, 'chapterTitle') ? stringifyEditorState(get(ref.referenceInfo, 'chapterTitle')) : undefined,
    has(ref.referenceInfo, 'articleTitle') ? stringifyEditorState(get(ref.referenceInfo, 'articleTitle')) : undefined,
    has(ref.referenceInfo, 'dataTitle') ? stringifyEditorState(get(ref.referenceInfo, 'dataTitle')) : undefined,
    has(ref.referenceInfo, 'source') ? stringifyEditorState(get(ref.referenceInfo, 'source')) : undefined,
    has(ref.referenceInfo, 'conferenceName')
      ? stringifyEditorState(get(ref.referenceInfo, 'conferenceName'))
      : undefined
  ]
    .filter(Boolean)
    .join('. ');
};

export const ReferenceCitationEditorPopup: React.FC<ReferenceCitationEditorPopupProps> = (props) => {
  const { editorView, node, onClose, onChange, anchorEl } = props;
  const refs = useSelector(getReferences);
  const [filteredRefs, setFilteredRefs] = useState<Reference[]>(refs);
  const [filterValue, setFilterValue] = useState<string>('');
  const [isReferenceDialogShown, setReferenceDialogShown] = useState<boolean>(false);
  const classes = useReferenceEditorStyles();
  const dispatch = useDispatch();

  const refId = node.attrs['refId'];
  const handleFilterChange = useCallback(
    (event) => {
      const filterValue = event.target['value'].toLowerCase();
      setFilterValue(event.target['value']);
      setFilteredRefs(
        refs.filter((ref) => {
          return getRefListItemText(ref).toLowerCase().indexOf(filterValue) >= 0 || refId === ref.id;
        })
      );
    },
    [refs, setFilterValue, refId]
  );
  const handleClick = useCallback(
    (event: SyntheticEvent) => {
      event.preventDefault();
      event.stopPropagation();
      const newRefId = (event.currentTarget as HTMLLIElement).dataset['refId'];
      if (newRefId !== refId) {
        onChange(refs.find(({ id }) => id === newRefId));
      } else {
        onChange(undefined);
      }
    },
    [onChange, refId, refs]
  );

  const handleAddReference = useCallback(
    (newReference) => {
      dispatch(manuscriptActions.addReferenceAction(newReference));
      setReferenceDialogShown(false);
      onChange(newReference);
    },
    [dispatch, onChange, setReferenceDialogShown]
  );

  const clearFilterField = useCallback(() => {
    setFilterValue('');
    setFilteredRefs(refs);
  }, [setFilterValue, setFilteredRefs, refs]);

  const openReferenceFormDialog = useCallback(() => {
    setReferenceDialogShown(true);
  }, []);

  const closeReferenceFormDialog = useCallback(() => {
    setReferenceDialogShown(false);
  }, []);

  if (!editorView) {
    return <></>;
  }

  return (
    <Popper open={true} anchorEl={anchorEl} style={{ zIndex: 1 }}>
      <Paper>
        <ClickAwayListener onClickAway={onClose} mouseEvent="onMouseUp">
          <div>
            <TextField
              classes={{ root: classes.filterField }}
              InputLabelProps={{ shrink: true }}
              label="Filter list"
              value={filterValue}
              variant="outlined"
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">
                    <ClearIcon
                      onClick={clearFilterField}
                      color="disabled"
                      classes={{ root: classes.clearFilterIcon }}
                    />
                  </InputAdornment>
                )
              }}
              onChange={handleFilterChange}
            />
            <ul className={classes.refSelectionList}>
              <li className={classes.refSelectionListItem} onClick={openReferenceFormDialog}>
                <AddIcon fontSize="small" classes={{ root: classes.addReferenceIcon }} /> Add Reference
              </li>
              {filteredRefs.map((ref) => (
                <li className={classes.refSelectionListItem} key={ref.id} data-ref-id={ref.id} onClick={handleClick}>
                  <div className={classes.refContent}>
                    <Interweave content={getRefListItemText(ref)} />
                  </div>
                  <div className={classes.refTick}>
                    <CheckCircleIcon color="primary" className={ref.id !== refId ? classes.hiddenIcon : ''} />
                  </div>
                </li>
              ))}
            </ul>
            {isReferenceDialogShown
              ? renderReferenceModal({
                  reference: undefined,
                  onAccept: handleAddReference,
                  onCancel: closeReferenceFormDialog,
                  onDelete: undefined
                })
              : undefined}
          </div>
        </ClickAwayListener>
      </Paper>
    </Popper>
  );
};
