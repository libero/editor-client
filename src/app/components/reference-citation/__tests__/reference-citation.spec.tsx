import React from 'react';
import { Node as ProsemirrorNode } from 'prosemirror-model';
import { EditorView } from 'prosemirror-view';
import { EditorState } from 'prosemirror-state';
import { ReplaceStep } from 'prosemirror-transform';

import { ReferenceCitationNodeView } from 'app/components/reference-citation';
import { createBodyState } from 'app/models/body';
import { Reference } from 'app/models/reference';

jest.mock('prosemirror-view');
jest.mock('app/components/reference-citation/reference-citation-editor-popup', () => ({
  ReferenceCitationEditorPopup: () => <div data-cmp="ReferenceCitationEditorPopup" />
}));

jest.mock('@material-ui/core/styles', () => {
  return {
    ThemeProvider: ({ children }) => <>{children}</>,
    createMuiTheme: jest.fn(),
    makeStyles: jest.requireActual('@material-ui/core/styles').makeStyles
  };
});

describe('Reference citation node view', () => {
  it('should render node', () => {
    const [state, node] = givenState();
    const view = givenEditorView(state);
    const nodeView = new ReferenceCitationNodeView(node, view, jest.fn());
    expect(nodeView.dom.innerHTML).toMatchSnapshot();
  });

  it('should open editor on click', () => {
    const parentContainer = document.createElement('div');
    parentContainer.id = 'view-parent-container';
    const [state, node] = givenState();
    const view = givenEditorView(state);
    const nodeView = new ReferenceCitationNodeView(node, view, jest.fn());
    parentContainer.appendChild(view.dom);
    nodeView.dom.dispatchEvent(new Event('click'));
    expect(parentContainer.innerHTML).toMatchSnapshot();
  });

  it('should update node markup', () => {
    const [state, node] = givenState();
    const view = givenEditorView(state);
    const nodeView = new ReferenceCitationNodeView(node, view, () => 1);
    const ref = new Reference();
    nodeView.handleChange(ref);
    expect(view.dispatch['mock'].calls[0][0].steps[0]).toBeInstanceOf(ReplaceStep);
  });

  it('should remove popup on close', () => {
    const parentContainer = document.createElement('div');
    parentContainer.id = 'view-parent-container';
    const [state, node] = givenState();
    const view = givenEditorView(state);
    const nodeView = new ReferenceCitationNodeView(node, view, jest.fn());
    parentContainer.appendChild(view.dom);
    nodeView.dom.dispatchEvent(new Event('click'));
    nodeView.close();
    expect(parentContainer.innerHTML).toBe(view.dom.outerHTML);
  });

  it('should ignore mutations', () => {
    const parentContainer = document.createElement('div');
    parentContainer.id = 'view-parent-container';
    const [state, node] = givenState();
    const view = givenEditorView(state);
    const nodeView = new ReferenceCitationNodeView(node, view, jest.fn());
    parentContainer.appendChild(view.dom);
    expect(nodeView.ignoreMutation()).toBe(true);
  });

  it('should events', () => {
    const parentContainer = document.createElement('div');
    parentContainer.id = 'view-parent-container';
    const [state, node] = givenState();
    const view = givenEditorView(state);
    const nodeView = new ReferenceCitationNodeView(node, view, jest.fn());
    parentContainer.appendChild(view.dom);
    expect(nodeView.stopEvent()).toBe(true);
  });
});

function givenState(): [EditorState, ProsemirrorNode] {
  let state = createBodyState(document.createElement('div'), '');
  const node = state.schema.nodes.figureCitation.createAndFill({ figIds: ['id1'] }, state.schema.text('test content'));
  state = state.apply(state.tr.insert(0, node));
  return [state, node];
}

function givenEditorView(state: EditorState): EditorView {
  const dom = document.createElement('div');
  dom.id = 'view-container';
  const view = new EditorView(dom, {
    state: null,
    dispatchTransaction: jest.fn()
  });
  view.dom = dom;
  view.state = state;
  view.dispatch = jest.fn();
  return view;
}
