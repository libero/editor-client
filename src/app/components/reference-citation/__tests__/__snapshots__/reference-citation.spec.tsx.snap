// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reference citation node view should open editor on click 1`] = `"<div id=\\"view-container\\"></div><div style=\\"position: absolute; z-index: 10;\\"><div data-cmp=\\"ReferenceCitationEditorPopup\\"></div></div>"`;

exports[`Reference citation node view should render node 1`] = `"???"`;
